web_src/js/features/repo-branch.js

```diff
@@ -1,6 +1,6 @@
 export function initRepoBranchButton() {
   $('.show-create-branch-modal.button').on('click', function () {
-    $('#create-branch-form')[0].action = $('#create-branch-form').data('base-action') + $(this).data('branch-from');
+    $('#create-branch-form')[0].action = $('#create-branch-form').data('base-action') + $(this).data('branch-from-urlcomponent');
```

Comment:
Maybe we can do PathEscapeSegments in JS, then we do not need the `data-branch-from-urlcomponent`

Reply:
If this is the only place we need it - no.

Reply:
This question is about the framework design.

Usually encoding/decoding should not cross different modules (otherwise double-encoding or non-encoding bugs may occur). So it's usually better to make every module handle encoding and decoding correctly.

If we do not introduce PathEscapeSegment now, maybe we can leave a comment to tell developers in future: if more JS modules need process paths, we should introduce a PathEscapeSegments for JS then.

Reply:
I understand, for this PR, I think what I'd like to do is to get people to think a bit more clearly about whether something is going in to an URL or whether it's getting presented.

At the moment what we currently believe is that you can always take the username of a user, reponame etc. and put it into an url directly. Similarly for Branch names etc.

I do not believe that is necessarily true - especially when thinking about how to deal with homoglyph attacks. By separating these two here, when we have a clearer idea of how to make urls and how to homoglyph protect we can merge the two again. 

Thus, my intention with data-branch-urlcomponent is to be deliberately ugly. It's not supposed to be the end.

---

As a note: 

Most of this PR is ugly as hell - there are hints at framework improvements, e.g. use url converter functions like `.Link()` `.HTMLURL()`. But these converter functions are all at the wrong level they need to be available at the template level/router level.

Reply:
(For example I think the use of setting.AppSubUrl and setting.AppUrl in models isn't great. If we can move the url generation up into the router level etc we can make this context dependent and make things like requests from [internal] get into internal based urls and [external] get external urls.)